File: core/src/main/resources/org/treblereel/gwt/three4g/renderers/webgl/WebGLProperties.js
import {WebGLProperty} from '../../closure/renderers/webgl/WebGLProperty.js';

class WebGLProperties {


	constructor() {
		/**
		 * @type {WeakMap<Object, WebGLProperty>}
		 */
		this.properties = new WeakMap();

	}

	/**
	 * @param {Object} object 
	 * @return {WebGLProperty}
	 */
	get( object ) {

		let map = this.properties.get( object );

		if ( map === undefined ) {

			map = {};
			this.properties.set( object, map );

		}

		return map;

	}

	/**
	 * 
	 * @param {Object} object 
	 */
	remove( object ) {

		this.properties.delete( object );

	}

	/**
	 * 
	 * @param {Object} object 
	 * @param {string} key 
	 * @param {*} value 
	 */
	update( object, key, value ) {

		this.properties.get( object )[ key ] = value;

	}

	dispose() {

		this.properties = new WeakMap();

	}

}


export { WebGLProperties };

File: core/src/main/java/org/treblereel/gwt/three4g/renderers/webgl/WebGLProperties.java
package org.treblereel.gwt.three4g.renderers.webgl;

import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsType;

@JsType(isNative = true, name = "THREE.WebGLProperties", namespace = JsPackage.GLOBAL)
public class WebGLProperties {
  public native void dispose();

  public native double get(double object);

  public native void remove(double object);

  public native double update(double object, double key, double value);
}
